File: Step1_dataIO/dcimg2tiff.py
from ast import Try
from os.path import basename
from numpy import ceil,log2,floor,amax
from tifffile import TiffWriter
from dcimg import *
from rich.progress import track

def dcimg2tiff(scrname,destfolder,thres):
	filename = basename(scrname)
	destname = destfolder+filename[:-6]
	print('[NOTE] Files will be saved in ',destname+'(*).tif')
	if filename[-6:]!='.dcimg':
		print('[ERR] File \"',filename,'\" should be in *.dcimg format.')
		return
	else:
		scrfile = DCIMGFile(scrname)
		print('========== Reading succeeded! ==========')
		tifnum = int(ceil(2**((log2(scrfile.shape[0])+log2(scrfile.shape[1])+log2(scrfile.shape[2]))-31)))
		pagenum = int(floor(2**(31-log2(scrfile.shape[1])-log2(scrfile.shape[2]))))-1
		print('File in shape ',scrfile.shape,', tif numbers: ',tifnum,', page numbers: ',pagenum)
		for tifind in range(tifnum):
			destname_ind = destname + '('+str(tifind+1)+').tif'
			with TiffWriter(destname_ind) as tif:
				for pageind in track(range(pagenum), description=filename+'('+str(tifind+1)+'):'):
					if pagenum*tifind+pageind<scrfile.shape[0]:
						maxv = amax(scrfile[pagenum*tifind+pageind,:,:])
						if maxv>thres:
							tif.write(scrfile[pagenum*tifind+pageind,:,:],contiguous=True)
					else:
						break



if __name__ == '__main__':
	scrfolder = 'E:\\XW20220812\\'
	# scrfolder = 'D:\\'
	# destfolder = 'I:\\XW20220620\\rawtif\\'
	# scrfolder = '\\\\143.215.226.127\\E\\XW20220810\\'
	destfolder = 'Z:\\Xuanwen\\FLFMuf\\ExpData\\XW20220812\\rawtif\\'
	scr_prefix = ['JK_414_','JK_414_speedtest_']
	for scr_pref in scr_prefix:
		for scrind in range(1,14):
			scrind_str = "%05d" % scrind
			scrname = scrfolder + scr_pref + scrind_str + '.dcimg'
			print(scrname)
			try:
				dcimg2tiff(scrname,destfolder,200)
			except:
				# raise
				print('Conversion error! Process stopped. [',scr_pref,', ',scrind,']')
				continue

File: Step1_dataIO/dcimg2tiff.m
function dcimg2tiff(scrnamelist,destfolder,varargin)
%DCIMG2TIFF Summary of this function goes here
%   Detailed explanation goes here

mod = py.importlib.import_module('dcimg');
py.importlib.reload(mod);
if exist(destfolder,'dir')==0
    mkdir(destfolder)
end

for scrind = 1:1:size(scrnamelist,1)
    [~,name,ext] = fileparts(scrnamelist(scrind));
    destname = [destfolder,'\',char(name)];
    if ext ~= ".dcimg"
        disp(['[ERR] File \"',char(name),char(ext),'\" should be in *.dcimg format.'])
        break
    else
        disp(['[NOTE] Files will be saved in ',destname,'(*).tif']);
        scrfile = py.dcimg.DCIMGFile(py.str(scrnamelist(scrind)));
        scrfile_shape = getProperty(scrfile,"shape");
        tifnum = round(ceil(2^((log2(getProperty(scrfile,"zsize"))+...
                                log2(getProperty(scrfile,"ysize"))+...
                                log2(getProperty(scrfile,"xsize")))-31)));
		pagenum = round(floor(2^(31-log2(getProperty(scrfile,"ysize"))-...
                                    log2(getProperty(scrfile,"xsize")))))-1;
        disp(['File in shape (',num2str(scrfile_shape),'), tif numbers: ',...
            num2str(tifnum),', page numbers: ',num2str(pagenum)])

        for tifind = 1:1:tifnum
            destname_ind = [destname, '(',num2str(tifind),').tif'];
            for pageind  = 1:1:pagenum
                if pageind ==1
                    t = Tiff(destname_ind,'w');
                else
                    t = Tiff(destname_ind,'a');
                end
                t.setTag('ImageLength', scrfile_shape(1));
                t.setTag('ImageWidth',  scrfile_shape(2));
                t.setTag('Photometric', Tiff.Photometric.MinIsBlack);
                t.setTag('Compression', Tiff.Compression.None);
                t.setTag('BitsPerSample', 16);
                t.setTag('SamplesPerPixel', 1);
                t.setTag('PlanarConfiguration', Tiff.PlanarConfiguration.Chunky);
                t.write( uint16(pyrun(['scrdata = pyscrfile[',num2str(pagenum*(tifind-1)+pageind),',:,:]'],...
                    "scrdata",pyscrfile=scrfile)) );
                t.close;
                disp([scrind,tifind,pageind])
            end
        end

    end
end
end

function propV = getProperty(dcimgfile,prop)
DCIMG_PROP = ["byte_depth","double";
              "bytes_per_img","double";
              "bytes_per_row","double";
              "deep_copy_enabled","";
              "dtype","";
              "file_size","double";
              "framestamps","double";
              "nfrms","double";
              "shape","cell";
              "timestamps","double";
              "xsize","double";
              "ysize","double";
              "zsize","double";
              "y0","double";
              "file_path","string";
              "first_4px_correction_enabled","";
              "binning","";
              "x0","double";
              "mma","double";
              "fmt_version","double";
              "mm",""];
if ismember(string(prop),DCIMG_PROP)
    [proprow,~] = find(DCIMG_PROP==string(prop));
    eval(['pypropV = dcimgfile.',char(prop),';'])
    eval(['propV = ',char(DCIMG_PROP(proprow,2)),'(pypropV);'])
    if prop=="shape"
        propV = [double(propV{2}),double(propV{3}),double(propV{1})];
    end
else
    disp([char(prop),' is not accessible!'])
end
end